### PR TITLE
P0896R4 changes to insert iterators

### PR DESCRIPTION
**Pull request by CaseyCarter**
# Description

When `defined(__cpp_lib_concepts)`, `(back_|front_|)insert_iterator` are default constructible, and have `ptrdiff_t` as difference type.

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

File: stl/inc/iterator [hunk 1]
```diff
@@ -25,12 +25,18 @@ class back_insert_iterator { // wrap pushes to back of container as output itera
...
     using iterator_category = output_iterator_tag;
     using value_type        = void;
-    using difference_type   = void;
     using pointer           = void;
     using reference         = void;
 
     using container_type = _Container;
 
+#ifdef __cpp_lib_concepts
+    using difference_type                     = ptrdiff_t;
```
Comment (StephanTLavavej) [inline stl/inc/iterator]: I would like to see a newline since there is no commonality between the typedef and the defaulted constructor. (Even if we didn't have align consecutive assignments). Ditto below.

File: stl/inc/iterator [hunk 2]
```diff
@@ -119,24 +131,28 @@ class insert_iterator { // wrap inserts into container as output iterator
 public:
     using iterator_category = output_iterator_tag;
     using value_type        = void;
-    using difference_type   = void;
     using pointer           = void;
     using reference         = void;
+    using container_type    = _Container;
```
Comment (StephanTLavavej) [inline stl/inc/iterator]: Why are you grouping `container_type` with the other typedefs in only this iterator? The change is fine, but should be done consistently.

File: stl/inc/iterator [hunk 3]
```diff
@@ -155,8 +171,8 @@ public:
     }
 
 protected:
-    _Container* container; // pointer to container
-    typename _Container::iterator iter; // iterator into container
+    _Container* container              = nullptr;
+    typename _Container::iterator iter = typename _Container::iterator();
```
Comment (StephanTLavavej) [inline stl/inc/iterator]: Even if you use copy-init, we should use empty braces for consistency with our new improved conventions. (I'm not sure if we can use direct-list-init; can we?)

File: stl/inc/iterator [hunk 4]
```diff
@@ -119,24 +131,28 @@ class insert_iterator { // wrap inserts into container as output iterator
...
     using value_type        = void;
-    using difference_type   = void;
     using pointer           = void;
     using reference         = void;
+    using container_type    = _Container;
 
-    using container_type = _Container;
+#ifdef __cpp_lib_concepts
+    using difference_type = ptrdiff_t;
+    insert_iterator()     = default;
```
Comment (miscco) [inline stl/inc/iterator]: You are missing constexpr/_CONSTEXPR20 here.

By the way #491 hisses at you mightily

File: stl/inc/iterator (hunk 2 above)
Comment (CaseyCarter) [reply stl/inc/iterator]: I missed reverting this one. I had many more non-functional cleanups here before realizing that they would conflict with #491 and backing them out.

File: stl/inc/iterator (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/iterator]: I'm mindlessly following the `type name = type();` style in the working draft which is intended to defend against Tim Song types (aggregates with a subobject that has an explicit default constructor). That style is unnecessary for "new" iterators since we fixed the `default_initializable` concept to require `T{}` to be valid. _`Cpp17DefaultConstructible`_ already required `T{}` for _`Cpp17ForwardIterator`_, I'll pretend that the handwavy front matter in [insert.iterators] actually requires `Container` to meet the container requirements and have a _`Cpp17ForwardIterator`_.

TLDR: Yes, we can direct-list-init or copy-list-init.

File: tests/std/tests/P0896R4_ranges_iterator_machinery/test.cpp [hunk 5]
```diff
@@ -2770,6 +2772,26 @@ namespace iter_ops {
     }
 } // namespace iter_ops
 
+namespace insert_iterators {
+    template <class Container>
+    constexpr bool test() {
+        using std::back_insert_iterator, std::front_insert_iterator, std::insert_iterator;
+        using std::default_initializable, std::iter_difference_t, std::ptrdiff_t, std::same_as;
+
+        STATIC_ASSERT(default_initializable<back_insert_iterator<Container>>);
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_iterator_machinery/test.cpp]: I believe we should also check `std::is_nothrow_default_constructible` given that these are explicitely `noexcept`

File: stl/inc/iterator (hunk 4 above)
Comment (CaseyCarter) [reply stl/inc/iterator]: This defaulted default constructor doesn't have an explicit `constexpr` in the working draft. I'm stomping on #491's context a bit, but I think I've managed to avoid stomping on the actual changed lines - resolving the conflicts should be trivial.

**Comment by CaseyCarter** [comment]
> 1. Any operation on a default constructed `*_insert_iterator` is an invalid operation. So I would suggest
> 
> ```c++
> ifdef __cpp_lib_concepts
> #if _ITERATOR_DEBUG_LEVEL >= 1
>         _STL_VERIFY(container, "cannot {dereference, increment} value-initialized  *_insert_iterator");
> #endif // _ITERATOR_DEBUG_LEVEL >= 1
> ifdef // __cpp_lib_concepts
> ```

Dereference and increment (since they have no effects) are well-defined for all three insert iterator types. `++++++++***++*++**x`, for example, is a perfectly valid lvalue that denotes `x`. The only invalid operations are assignment from lvalue/rvalue `Container::value_type`. Those operations immediately dereference `this->container` which is known to be a `nullptr` for default-initialized insert iterators. We conventionally don't guard operations that will "safely" be terminated due to null pointer dereference unless it's as part of a larger set of consistent checks. 

> 2. It feels rather inconsistent that we require the -real- constructors to be `explicit `and happily let the user default construct them anyway. Adding `explicit` requires an LWG-issue right?

WG21 traditionally annotates single-argument converting constructors as `explicit` to prevent inadvertent implicit conversions. There have been some rumblings recently about making multiple-argument constructors `explicit`, but generally speaking default/copy/move constructors _should_ be implicit.

> 3. If we are already looking at concepts should they actually be constrained by something along those lines
> 
> ```c++
> ifdef __cpp_lib_concepts
>     requires requires(_Container& __c) { __c.push_back(); }
> ifdef // __cpp_lib_concepts
> ```
> 
> Another LWG issue maybe?

This probably needs a design paper. For Ranges, Eric and I decided that our goal was to conceptualize the algorithms and we deliberately put other kinds of concept constraints out of scope. We didn't want standardize ad hoc concepts ([not that we didn't have some](https://github.com/CaseyCarter/cmcstl2/blob/master/include/stl2/detail/iterator/insert_iterators.hpp#L47-L51)) that might later interfere with "proper" conceptualization of containers

File: stl/inc/iterator [hunk 6]
```diff
@@ -119,24 +133,30 @@ class insert_iterator { // wrap inserts into container as output iterator
...
-        const typename _Container::value_type& _Val) { // insert into container and increment stored iterator
+    insert_iterator& operator=(const typename _Container::value_type& _Val) {
+        // insert into container and increment stored iterator
         iter = container->insert(iter, _Val);
         ++iter;
         return *this;
     }
 
     insert_iterator& operator=(typename _Container::value_type&& _Val) { // push value into container
-
```
Comment (CaseyCarter) [inline stl/inc/iterator]: Backout the above two changes to avoid conflicts with #491.

File: stl/inc/iterator [hunk 7]
```diff
@@ -155,8 +176,8 @@ public:
     }
 
 protected:
-    _Container* container; // pointer to container
-    typename _Container::iterator iter; // iterator into container
+    _Container* container              = nullptr;
+    typename _Container::iterator iter = {};
```
Comment (StephanTLavavej) [inline stl/inc/iterator]:
```suggestion
    typename _Container::iterator iter{};
```

Is this friendlier to the Cpp17DefaultConstructible requirements? The Standard depicts `ranges::iterator_t<Container> iter = ranges::iterator_t<Container>();` which is explicit default construction of a temporary followed by copy-init copy-construction. Saying `= {}` is implicit default construction. Saying `{}` is explicit default construction.

(I no longer pretend to understand C++ initialization. I guess I traded that for the relative simplicity of floating-point.)

Comment (CaseyCarter) [reply stl/inc/iterator]: The form in the standard - `T t = T();` - is the most-generic form we have to express value-initialization. (There's no copy/move here since C++17: this simply initializes `t` with the prvalue `T()` per [dcl.init]/17.6.1.) We needed that fully generic form here before LWG-3149 fixed the `default_initializable` concept to require `T{}` to be valid.

I used copy-list init here because I had the impression you disliked mixing direct and copy init, not for any technical reason.

File: stl/inc/iterator [hunk 8]
```diff
@@ -155,8 +176,8 @@ public:
     }
 
 protected:
-    _Container* container; // pointer to container
-    typename _Container::iterator iter; // iterator into container
+    _Container* container              = nullptr;
```
Comment (StephanTLavavej) [inline stl/inc/iterator]: This is failing clang-format; consider enabling format-on-save in your editor. Everything else looks good.

**Comment by miscco** [comment]
Plesse do not trouble you too much with merge conflicts with other PRs. It is totally fine to resolve those in the other PR.

I will hiss and wail though but that is fine

File: stl/inc/iterator (hunk 7 above)
Comment (StephanTLavavej) [reply stl/inc/iterator]: I very mildly prefer consistency but copy-init of scalars doesn't really bother me.

File: stl/inc/iterator [hunk 9]
```diff
@@ -119,12 +133,19 @@ class insert_iterator { // wrap inserts into container as output iterator
...
-    using difference_type   = void;
     using pointer           = void;
     using reference         = void;
 
     using container_type = _Container;
 
+#ifdef __cpp_lib_concepts
+    using difference_type = ptrdiff_t;
+
+    insert_iterator() = default;
```
Comment (barcharcraz) [inline stl/inc/iterator]: out of curiosity: Why isn't this one constexpr (in the standard)

Comment (CaseyCarter) [reply stl/inc/iterator]: I have absolutely no idea. (To be honest, I don't understand why we ever put an explicit `constexpr` on a defaulted default constructor.) These default constructors came from P0896, which crossed with P1032 that added `constexpr` to all of the members of the insert iterators, and apparently the editors decided to put `constexpr` on some subset of the default constructors for reasons not apparent to me.
